sdks/java/io/hdfs/src/main/java/org/apache/beam/sdk/io/hdfs/HDFSFileSink.java

```diff
@@ -0,0 +1,280 @@
+import org.apache.hadoop.mapreduce.task.TaskAttemptContextImpl;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A {@code Sink} for writing records to a Hadoop filesystem using a Hadoop file-based output
+ * format.
+ *
+ * @param <K> The type of keys to be written to the sink.
+ * @param <V> The type of values to be written to the sink.
+ */
+public class HDFSFileSink<K, V> extends Sink<KV<K, V>> {
+
+  private static final String jtIdentifier = "scio_job";
+
+  protected final String path;
+  protected final Class<? extends FileOutputFormat<K, V>> formatClass;
+
+  // workaround to make Configuration serializable
+  private final Map<String, String> map;
+
+  public HDFSFileSink(String path, Class<? extends FileOutputFormat<K, V>> formatClass) {
+    this.path = path;
+    this.formatClass = formatClass;
+    this.map = Maps.newHashMap();
+  }
+
+  public HDFSFileSink(String path, Class<? extends FileOutputFormat<K, V>> formatClass,
+                      Configuration conf) {
+    this(path, formatClass);
+    // serialize conf to map
+    for (Map.Entry<String, String> entry : conf) {
+      map.put(entry.getKey(), entry.getValue());
+    }
+  }
+
+  @Override
+  public void validate(PipelineOptions options) {
+    try {
+      Job job = jobInstance();
+      FileSystem fs = FileSystem.get(job.getConfiguration());
+      Preconditions.checkState(
+              !fs.exists(new Path(path)),
+              "Output path " + path + " already exists");
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public Sink.WriteOperation<KV<K, V>, ?> createWriteOperation(PipelineOptions options) {
+    return new HDFSWriteOperation<>(this, path, formatClass);
+  }
+
+  private Job jobInstance() throws IOException {
+    Job job = Job.getInstance();
+    // deserialize map to conf
+    Configuration conf = job.getConfiguration();
+    for (Map.Entry<String, String> entry : map.entrySet()) {
+      conf.set(entry.getKey(), entry.getValue());
+    }
+    return job;
+  }
+
+  // =======================================================================
+  // WriteOperation
+  // =======================================================================
+
+  /** {{@link org.apache.beam.sdk.io.Sink.WriteOperation}} for HDFS. */
+  public static class HDFSWriteOperation<K, V> extends WriteOperation<KV<K, V>, String> {
+
+    private final Sink<KV<K, V>> sink;
+    protected final String path;
+    protected final Class<? extends FileOutputFormat<K, V>> formatClass;
+
+    // unique job ID for this sink
+    private final int jobId;
+
+    public HDFSWriteOperation(Sink<KV<K, V>> sink,
+                              String path,
+                              Class<? extends FileOutputFormat<K, V>> formatClass) {
+      this.sink = sink;
+      this.path = path;
+      this.formatClass = formatClass;
+      this.jobId = (int) (System.currentTimeMillis() / 1000);
```

Comment:
Updated it so that each `HDFSSink` gets a unique `JobID` with `currentTimeMillis` + random integer.

We never submit the job as a M/R or YARN application so JobID collision shouldn't be a concern here. Each writer will create a temporary output directory of `${outputdir}/_temporary/_${taskid}` where `taskid` includes job id, task type and hash of uId.